pennylane/transforms/qcut.py

```diff
@@ -210,3 +212,80 @@ def tape_to_graph(tape: QuantumTape) -> MultiDiGraph:
             order += 1
 
     return graph
+
+
+def fragment_graph(graph: MultiDiGraph) -> Tuple[Tuple[MultiDiGraph], MultiDiGraph]:
+    """
+    Fragments a cut graph into a collection of subgraphs as well as returning
+    the communication/quotient graph.
+
+    Args:
+        graph (MultiDiGraph): directed multigraph containing measure and prepare
+        nodes at cut locations
+
+    Returns:
+        subgraphs, communication_graph (Tuple[Tuple[MultiDiGraph], MultiDiGraph]):
+        the subgraphs of the cut graph and the communication graph where each
+        node represents a fragment and edges denote the flow of qubits between
+        fragments
+
+    **Example**
+
+    Consider the following circuit with the manually-placed wire cuts:
+
+    .. code-block:: python
+
+        from pennylane.transforms import qcut
+
+        wire_cut_0 = qml.WireCut(wires=0)
+        wire_cut_1 = qml.WireCut(wires=1)
+        multi_wire_cut = qml.WireCut(wires=[0, 1])
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.4, wires=0)
+            qml.apply(wire_cut_0)
+            qml.RY(0.5, wires=0)
+            qml.apply(wire_cut_1)
+            qml.CNOT(wires=[0, 1])
+            qml.apply(multi_wire_cut)
+            qml.RZ(0.6, wires=1)
+            qml.expval(qml.PauliZ(0))
+
+    We can find the corresponding graph, remove all the wire cut nodes, and
+    find the subgraphs and communication graph by using:
+
+    >>> graph = qcut.tape_to_graph(tape)
+    >>> qcut.replace_wire_cut_nodes(graph)
+    >>> qcut.fragment_graph(graph)
+    ((<networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b2311940>,
+      <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b2311c10>,
+      <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b23e2820>,
+      <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b23e27f0>),
+     <networkx.classes.multidigraph.MultiDiGraph object at 0x7fb3b23e26a0>)
+    """
+
+    edges = list(graph.edges)
+    cut_edges = []
+
+    for node1, node2, _ in edges:
+        if isinstance(node1, MeasureNode):
+            assert isinstance(node2, PrepareNode)
+            cut_edges.append((node1, node2))
+            graph.remove_edge(node1, node2)
```

Comment:
I would make a (deep) copy of `graph` beforehand to avoid of risk of unwanted side effects, as long as the graph is not too large, which is the case for us right? Adding edges back in could change the internal order of how things are stored, which is always a risk.

Reply:
Thanks for the suggestions, I've gone with the deep copy approach and added a unit test for the communication graph over repeated applications of `fragment_graph(graph)`